v4/core/request_builder.go

```diff
@@ -364,6 +365,12 @@ func (requestBuilder *RequestBuilder) Build() (req *http.Request, err error) {
 		req = req.WithContext(requestBuilder.ctx)
 	}
 
+	// Provide a way for the user to print out the request, for debugging purposes
```

Comment:
Okay, I'll look at moving the code out of that method 👍

> I think we should end up calling the DumpRequestOut() method ONLY if it's requested by the user
Is logging it at the highest level (Debug) not a conditional in and of itself? A user would have to explicitly request logging at that level to see the dump. Unless the dump constitutes much of a performance hit, I don't really see the use in adding yet another conditional on top of that. What do you think?

Reply:
My main concern is that I feel like we should not do anything that could possibly affect the request if we're not going to do any debug logging.   In the case of logging the request details, the DumpRequestOut() method has to consume the request body then set a new io.ReadCloser on the request so that the normal request processing can still consume the request body bytes. In the case of a simple request body that's relatively small, that's not really a problem.  But if a request is using a large streamed JSON document (think Cloudant and their operations that stream JSON request body content, or really any large request body), then the act of just calling DumpRequestOut() means we have to consume all the streamed bytes (which could be several MBs in the Cloudant situation), display the debug output, then wrap the consumed bytes in another io.ReadCloser and set that on the request, thus reading and copying all the request body bytes.

I think instead of my suggestion to use an env variable, we could instead just base the decision on whether debug logging is enabled or not.   Currently, the Logger interface doesn't support this, but I think we could just add the following methods to the Logger interface to facilitate that:
```
IsDebugEnabled() bool
IsInfoEnabled() bool
IsWarnEnabled() bool
IsErrorEnabled() bool
```
Then implement them in the SDKLoggerImpl type like this:
```
func (l *SDKLoggerImpl) IsDebugEnabled() bool {
	return l.logLevel >= LevelDebug
}
func (l *SDKLoggerImpl) IsInfoEnabled() bool {
	return l.logLevel >= LevelInfo
}
...
```
Essentially add one "IsXXXEnabled()" for each log level.

Then you could surround the call to DumpRequestOut() with:
```
if GetLogger().IsDebugEnabled() {
    // log the request details
}
```

What do you think?

Reply:
> Unless the dump constitutes much of a performance hit

...

> thus reading and copying all the request body bytes

Point well made 🙂 I agree, and I like the approach that you've proposed